### PR TITLE
new:usr:#14 Support partial cubes

### PR DESCRIPTION
**Pull request by amoghmargoor**

### CODE REVIEW

**Comment by vrajat** [comment]
I need to step through the `FilterAggStarRule`. Did you use `AggStarRule` for inspiration ? I have to step through this to get it. Steps seem fine at a high level. I dont understand if all conditions are handled.

Tests are light though. You should start a new suite for filtered cubes. Let the current suite and sql files be. I can add some tests as I review.

**Comment by amoghmargoor** [comment]
yes it is inspired by AggStarRule.  I tried reducing the tree to what AggStarRule would handle...but that reduction was pretty heavy and i created a new rule for it. I have added new tests in LatticeTest and created new suite LatticeWithFilter. Will check those in with few fixes in about 15-20 mins.

**Comment by amoghmargoor** [comment]
@vrajat Have added the tests. There is one issue which i am currently fixing. When group by and filter are specified on same dimension then optimization is not happening. May be due to difference in tree .. I am fixing that and test for that is  @Ignored

**Comment by amoghmargoor** [comment]
@vrajat I have fixed that issue and new rule has been added for it. Only thing left is to make Lattice extension in Quark to get rid of all calcite changes for this PR.

File: optimizer/src/test/java/com/qubole/quark/planner/test/LatticeWithFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  public void testStoreCubeWithSatisfiedFilter() throws QuarkException, SQLException {
+      String sql = "select d_year, d_qoy, sum(ss_sales_price) " +
+          " from tpcds.store_sales join tpcds.date_dim on ss_sold_date_sk = d_date_sk " +
+          " join tpcds.customer_demographics on ss_cdemo_sk = cd_demo_sk " +
+          "where cd_gender = 'M' group by d_year, d_qoy";
+    QuarkTestUtil.checkParsedSql(
+        sql,
+        info,
+        "SELECT D_YEAR, D_QOY, SUM(SUM_SALES_PRICE)" +
+            " FROM TPCDS.STORE_SALES_CUBE WHERE CD_GENDER = 'M'" +
```
Comment (vrajat) [inline optimizer/src/test/java/com/qubole/quark/planner/test/LatticeWithFilterTest.java]: Shouldn't the filter `CD_GENDER = 'M'` be removed here ?

**Comment by vrajat** [comment]
I started a [suite](https://github.com/vrajat/quark/commit/a3a5efd737d04d7a37684ac68561fce2a2e5c1c8) to test realistic scenarios. I am not able to get Quark to use the cube in a sanity test. Can you eyeball the definition ? I cant find anything wrong.

**Comment by amoghmargoor** [comment]
@vrajat took a quick look at the suit. It points to schema factory of LatticeWithFilterTest. Instead it should point to PartialCubeTest.

File: optimizer/src/test/java/com/qubole/quark/planner/test/LatticeWithFilterTest.java (hunk 1 above)
Comment (amoghmargoor) [inline optimizer/src/test/java/com/qubole/quark/planner/test/LatticeWithFilterTest.java]: yeah ... haven't done that optimization yet... will do it in followup commit.

**Comment by vrajat** [comment]
Gah!

**Comment by vrajat** [comment]
Now getting a real error in the suite:
```
java.lang.AssertionError: Internal error: Error while applying rule FilterAggStarRule:FilterOnGroupSet, args [rel#12355:LogicalFilter.NONE.[](input=rel#12354:Subset#11.NONE.[],condition=AND(=(CAST($0):INTEGER, 2007), =(CAST($1):INTEGER, 1), =(CAST($2):INTEGER, 1))), rel#12985:LogicalAggregate.NONE.[](input=rel#12350:Subset#9.NONE.[],group={29, 31, 32, 52},EXPR$4=SUM($13)), rel#12349:LogicalProject.NONE.[](input=rel#12348:Subset#8.NONE.[],SS_SOLD_DATE_SK=$0,SS_SOLD_TIME_SK=$1,SS_ITEM_SK=$2,SS_CUSTOMER_SK=$3,SS_CDEMO_SK=$4,SS_HDEMO_SK=$5,SS_ADDR_SK=$6,SS_STORE_SK=$7,SS_PROMO_SK=$8,SS_TICKET_NUMBER=$9,SS_QUANTITY=$10,SS_WHOLESALE_COST=$11,SS_LIST_PRICE=$12,SS_SALES_PRICE=$13,SS_EXT_DISCOUNT_AMT=$14,SS_EXT_SALES_PRICE=$15,SS_EXT_WHOLESALE_COST=$16,SS_EXT_LIST_PRICE=$17,SS_EXT_TAX=$18,SS_COUPON_AMT=$19,SS_NET_PAID=$20,SS_NET_PAID_INC_TAX=$21,SS_NET_PROFIT=$22,D_DATE_SK=$63,D_DATE_ID=$64,D_DATE=$65,D_MONTH_SEQ=$66,D_WEEK_SEQ=$67,D_QUARTER_SEQ=$68,D_YEAR=$69,D_DOW=$70,D_MOY=$71,D_DOM=$72,D_QOY=$73,D_FY_YEAR=$74,D_FY_QUARTER_SEQ=$75,D_FY_WEEK_SEQ=$76,D_DAY_NAME=$77,D_QUARTER_NAME=$78,D_HOLIDAY=$79,D_WEEKEND=$80,D_FOLLOWING_HOLIDAY=$81,D_FIRST_DOM=$82,D_LAST_DOM=$83,D_SAME_DAY_LY=$84,D_SAME_DAY_LQ=$85,D_CURRENT_DAY=$86,D_CURRENT_WEEK=$87,D_CURRENT_MONTH=$88,D_CURRENT_QUARTER=$89,D_CURRENT_YEAR=$90,CD_DEMO_SK=$91,CD_GENDER=$92,CD_MARITAL_STATUS=$93,CD_EDUCATION_STATUS=$94,CD_PURCHASE_ESTIMATE=$95,CD_CREDIT_RATING=$96,CD_DEP_COUNT=$97,CD_DEP_EMPLOYED_COUNT=$98,CD_DEP_COLLEGE_COUNT=$99), rel#12384:StarTableScan.NONE.[](table=[QUARK_METADATA, store_sales_cube_partial])]
	at org.apache.calcite.rel.core.Filter.<init>(Filter.java:73)
	at org.apache.calcite.rel.logical.LogicalFilter.<init>(LogicalFilter.java:60)
	at org.apache.calcite.rel.logical.LogicalFilter.create(LogicalFilter.java:94)
	at com.qubole.quark.planner.FilterAggStarRule.constructTileRel(FilterAggStarRule.java:215)
	at com.qubole.quark.planner.FilterAggStarRule.apply(FilterAggStarRule.java:180)
	at com.qubole.quark.planner.FilterAggStarRule$1.onMatch(FilterAggStarRule.java:88)
	at org.apache.calcite.plan.volcano.VolcanoRuleCall.onMatch(VolcanoRuleCall.java:228)
	at org.apache.calcite.plan.volcano.VolcanoPlanner.findBestExp(VolcanoPlanner.java:826)
	at com.qubole.quark.sql.SqlWorker$EnumerableProgram.run(SqlWorker.java:204)
	at org.apache.calcite.tools.Programs$SequenceProgram.run(Programs.java:321)
	at org.apache.calcite.prepare.PlannerImpl.transform(PlannerImpl.java:273)
	at com.qubole.quark.sql.handlers.EnumerableSqlHandler.convert(EnumerableSqlHandler.java:54)
	at com.qubole.quark.sql.handlers.EnumerableSqlHandler.convert(EnumerableSqlHandler.java:33)
	at com.qubole.quark.sql.SqlWorker.parse(SqlWorker.java:240)
	at com.qubole.quark.planner.Parser.parseInternal(Parser.java:104)
	at com.qubole.quark.planner.Parser.parse(Parser.java:141)
	at com.qubole.quark.planner.test.utilities.QuarkTestUtil.getParsedSql(QuarkTestUtil.java:57)
	at com.qubole.quark.planner.test.utilities.QuarkTestUtil.checkParsedSql(QuarkTestUtil.java:62)
	at com.qubole.quark.planner.test.PartialCubeTest.storeFilterOnGroupDimensionWithSatisfiedFilter(PartialCubeTest.java:125)
```

**Comment by amoghmargoor** [comment]
@vrajat  Ahh!! that's bad ... creation of filter accepts only flattened condition...have fixed it ...can you try now ?

**Comment by vrajat** [comment]
Worked

**Comment by vrajat** [comment]
I keep going back & forth on this. Should the filters be on mandatory columns ? I know you dont have any validations. I continue to be confused about being strict about it.

**Comment by amoghmargoor** [comment]
Filter on mandatory dimension is required only if user wants to do incremental build. If user wants to create cube for 2 years say 2007 and 2008 without any incremental build then such restriction is not required. So to allow more use cases it is better not to enforce such restriction.

**Comment by amoghmargoor** [comment]
For some test queries, as tiles with exact dimensions are not available it picks up tiles that contain the required dimension as a subset and do rollup on it. As there could be many such tiles, one of them is picked. Now picking that tile is not deterministic and differs between java7 and java8. Hence modified the tests to take care of it.

**Pull request merged**